File: Caso2/TypeScript/Back/estructuras.ts
/**
 * Estructuras de datos propias para CentralAscensores.
 *
 * Nada de Array como almacenamiento de estas tres estructuras: cada una
 * se arma con nodos enlazados por punteros (this.siguiente), que es lo
 * que permite O(1) en los extremos correctos.
 */

class Nodo<T> {
  dato: T;
  siguiente: Nodo<T> | null = null;
  constructor(dato: T) {
    this.dato = dato;
  }
}

/** Lista enlazada simple con cabeza y cola: altas en cualquier momento,
 * sin orden de llegada ni de apilamiento, recorrido completo para reportes. */
export class Lista<T> {
  private cabeza: Nodo<T> | null = null;
  private cola: Nodo<T> | null = null;
  private tamano = 0;

  agregar(dato: T): void {
    const nodo = new Nodo(dato);
    if (this.cabeza === null) {
      this.cabeza = nodo;
      this.cola = nodo;
    } else {
      this.cola!.siguiente = nodo;
      this.cola = nodo;
    }
    this.tamano++;
  }

  buscar(predicado: (dato: T) => boolean): T | null {
    let actual = this.cabeza;
    while (actual !== null) {
      if (predicado(actual.dato)) return actual.dato;
      actual = actual.siguiente;
    }
    return null;
  }

  *recorrer(): Generator<T> {
    let actual = this.cabeza;
    while (actual !== null) {
      yield actual.dato;
      actual = actual.siguiente;
    }
  }

  get length(): number {
    return this.tamano;
  }
}

/** Cola FIFO: encolar/desencolar en O(1), desencolar solo mueve la cabeza. */
export class Cola<T> {
  private cabeza: Nodo<T> | null = null;
  private cola: Nodo<T> | null = null;
  private tamano = 0;

  encolar(dato: T): void {
    const nodo = new Nodo(dato);
    if (this.cola === null) {
      this.cabeza = nodo;
      this.cola = nodo;
    } else {
      this.cola.siguiente = nodo;
      this.cola = nodo;
    }
    this.tamano++;
  }

  desencolar(): T {
    if (this.cabeza === null) throw new Error("cola vacia");
    const nodo = this.cabeza;
    this.cabeza = nodo.siguiente;
    if (this.cabeza === null) this.cola = null;
    this.tamano--;
    return nodo.dato;
  }

  vacia(): boolean {
    return this.cabeza === null;
  }

  get length(): number {
    return this.tamano;
  }
}

/** Pila LIFO: push/pop/tope en O(1), solo se toca la cabeza. */
export class Pila<T> {
  private tope_: Nodo<T> | null = null;
  private tamano = 0;

  apilar(dato: T): void {
    const nodo = new Nodo(dato);
    nodo.siguiente = this.tope_;
    this.tope_ = nodo;
    this.tamano++;
  }

  desapilar(): T {
    if (this.tope_ === null) throw new Error("pila vacia");
    const nodo = this.tope_;
    this.tope_ = nodo.siguiente;
    this.tamano--;
    return nodo.dato;
  }

  tope(): T {
    if (this.tope_ === null) throw new Error("pila vacia");
    return this.tope_.dato;
  }

  vacia(): boolean {
    return this.tope_ === null;
  }

  get length(): number {
    return this.tamano;
  }
}

File: Caso2/Python/Back/estructuras.py
"""
Estructuras de datos propias para CentralAscensores.

No se usa list, collections.deque ni ninguna coleccion estandar como
almacenamiento de estas tres estructuras: cada una se construye con nodos
enlazados por punteros, que es lo que permite que insertar/quitar en los
extremos correctos sea O(1).
"""

from __future__ import annotations
from typing import Any, Callable, Optional


class _Nodo:
    __slots__ = ("dato", "siguiente")

    def __init__(self, dato: Any):
        self.dato = dato
        self.siguiente: Optional["_Nodo"] = None


class Lista:
    """Lista enlazada simple con cabeza y cola.

    Se usa para el parque de ascensores: hay altas en cualquier momento,
    no hay orden de llegada (no es FIFO) ni de apilamiento (no es LIFO),
    y hay que recorrer todo para los reportes -> ni cola ni pila encajan,
    una lista si.
    """

    def __init__(self):
        self._cabeza: Optional[_Nodo] = None
        self._cola: Optional[_Nodo] = None
        self._tamano = 0

    def agregar(self, dato: Any) -> None:
        nodo = _Nodo(dato)
        if self._cabeza is None:
            self._cabeza = nodo
            self._cola = nodo
        else:
            self._cola.siguiente = nodo
            self._cola = nodo
        self._tamano += 1

    def buscar(self, predicado: Callable[[Any], bool]) -> Optional[Any]:
        actual = self._cabeza
        while actual is not None:
            if predicado(actual.dato):
                return actual.dato
            actual = actual.siguiente
        return None

    def recorrer(self):
        actual = self._cabeza
        while actual is not None:
            yield actual.dato
            actual = actual.siguiente

    def __len__(self) -> int:
        return self._tamano


class Cola:
    """Cola FIFO con cabeza y cola (encolar/desencolar en O(1)).

    Desencolar solo mueve el puntero de cabeza: no recorre ni desplaza
    los elementos restantes.
    """

    def __init__(self):
        self._cabeza: Optional[_Nodo] = None
        self._cola: Optional[_Nodo] = None
        self._tamano = 0

    def encolar(self, dato: Any) -> None:
        nodo = _Nodo(dato)
        if self._cola is None:
            self._cabeza = nodo
            self._cola = nodo
        else:
            self._cola.siguiente = nodo
            self._cola = nodo
        self._tamano += 1

    def desencolar(self) -> Any:
        if self._cabeza is None:
            raise IndexError("cola vacia")
        nodo = self._cabeza
        self._cabeza = nodo.siguiente
        if self._cabeza is None:
            self._cola = None
        self._tamano -= 1
        return nodo.dato

    def vacia(self) -> bool:
        return self._cabeza is None

    def __len__(self) -> int:
        return self._tamano


class Pila:
    """Pila LIFO con push/pop/tope en O(1) (solo se toca la cabeza)."""

    def __init__(self):
        self._tope: Optional[_Nodo] = None
        self._tamano = 0

    def apilar(self, dato: Any) -> None:
        nodo = _Nodo(dato)
        nodo.siguiente = self._tope
        self._tope = nodo
        self._tamano += 1

    def desapilar(self) -> Any:
        if self._tope is None:
            raise IndexError("pila vacia")
        nodo = self._tope
        self._tope = nodo.siguiente
        self._tamano -= 1
        return nodo.dato

    def tope(self) -> Any:
        if self._tope is None:
            raise IndexError("pila vacia")
        return self._tope.dato

    def vacia(self) -> bool:
        return self._tope is None

    def __len__(self) -> int:
        return self._tamano
